File: src/main/java/catwalks/item/crafting/RecipeDecorationRepair.java
package catwalks.item.crafting;

import net.minecraft.inventory.InventoryCrafting;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.item.crafting.IRecipe;
import net.minecraft.world.World;

public class RecipeDecorationRepair implements IRecipe {

	Item targetItem;
	
	public Item getTargetItem() {
		return targetItem;
	}
	
	public RecipeDecorationRepair(Item item) {
		targetItem = item;
	}
	
	/**
     * Used to check if a recipe matches current crafting inventory
     */
    public boolean matches(InventoryCrafting inv, World worldIn) {
        int count = 0;
        
        for (int i = 0; i < inv.getSizeInventory(); ++i) {
            ItemStack itemstack = inv.getStackInSlot(i);

            if (itemstack != null) {
            	if(itemstack.getItem() != targetItem)
            		return false;
            	count++;
            }
        }
        
        return count > 1;
    }

    /**
     * Returns an Item that is the result of this recipe
     */
    public ItemStack getCraftingResult(InventoryCrafting inv) {
        
    	ItemStack stack = null;
    	
    	for (int i = 0; i < inv.getSizeInventory(); ++i) {
            ItemStack itemstack = inv.getStackInSlot(i);

            if (itemstack != null) {
            	if(itemstack.getItem() == targetItem) {
            		if(stack == null) {
            			stack = itemstack.copy();
            		} else {
            			int usesNeeded = stack.getItemDamage();
            			int usesAvailable = itemstack.getMaxDamage() - itemstack.getItemDamage();
            			int usesToTake = usesNeeded > usesAvailable ? usesAvailable : usesNeeded;
            			stack.setItemDamage(stack.getItemDamage() - usesToTake);
            		}
            	}
            }
        }

        return stack;
    }

    /**
     * Returns the size of the recipe area
     */
    public int getRecipeSize() {
        return 4;
    }

    public ItemStack getRecipeOutput() {
        return null;
    }

    public ItemStack[] getRemainingItems(InventoryCrafting inv) {
        ItemStack[] aitemstack = new ItemStack[inv.getSizeInventory()];

        ItemStack stack = null;
    	
    	for (int i = 0; i < inv.getSizeInventory(); ++i) {
            ItemStack itemstack = inv.getStackInSlot(i);

            if (itemstack != null) {
            	if(itemstack.getItem() == targetItem) {
            		if(stack == null) {
            			stack = itemstack.copy();
            		} else {
            			int usesNeeded = stack.getItemDamage();
            			int usesAvailable = itemstack.getMaxDamage() - itemstack.getItemDamage();
            			int usesToTake = usesNeeded > usesAvailable ? usesAvailable : usesNeeded;
            			
            			stack.setItemDamage(stack.getItemDamage() - usesToTake);
            			
            			if(usesNeeded < usesAvailable) {
            				itemstack = itemstack.copy();
            				itemstack.setItemDamage(itemstack.getItemDamage()+usesNeeded);
            				aitemstack[i] = itemstack;
            			}
            		}
            	}
            }
        }

        return aitemstack;
    }

}

File: src/main/java/catwalks/item/crafting/RecipeDecorationRepair.kt
package catwalks.item.crafting

import net.minecraft.inventory.InventoryCrafting
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.item.crafting.IRecipe
import net.minecraft.world.World

class RecipeDecorationRepair(item: Item) : IRecipe {

    var targetItem: Item
        internal set

    init {
        targetItem = item
    }

    /**
     * Used to check if a recipe matches current crafting inventory
     */
    override fun matches(inv: InventoryCrafting, worldIn: World): Boolean {
        var count = 0

        for (i in 0..inv.sizeInventory - 1) {
            val itemstack = inv.getStackInSlot(i)

            if (itemstack != null) {
                if (itemstack.item !== targetItem)
                    return false
                count++
            }
        }

        return count > 1
    }

    /**
     * Returns an Item that is the result of this recipe
     */
    override fun getCraftingResult(inv: InventoryCrafting): ItemStack? {

        var stack: ItemStack? = null

        for (i in 0..inv.sizeInventory - 1) {
            val itemstack = inv.getStackInSlot(i)

            if (itemstack != null) {
                if (itemstack.item === targetItem) {
                    if (stack == null) {
                        stack = itemstack.copy()
                    } else {
                        val usesNeeded = stack.itemDamage
                        val usesAvailable = itemstack.maxDamage - itemstack.itemDamage
                        val usesToTake = if (usesNeeded > usesAvailable) usesAvailable else usesNeeded
                        stack.itemDamage = stack.itemDamage - usesToTake
                    }
                }
            }
        }

        return stack
    }

    /**
     * Returns the size of the recipe area
     */
    override fun getRecipeSize(): Int {
        return 4
    }

    override fun getRecipeOutput(): ItemStack? {
        return null
    }

    override fun getRemainingItems(inv: InventoryCrafting): Array<ItemStack?> {
        val aitemstack = arrayOfNulls<ItemStack>(inv.sizeInventory)

        var stack: ItemStack? = null

        for (i in 0..inv.sizeInventory - 1) {
            var itemstack = inv.getStackInSlot(i)

            if (itemstack != null) {
                if (itemstack.item === targetItem) {
                    if (stack == null) {
                        stack = itemstack.copy()
                    } else {
                        val usesNeeded = stack.itemDamage
                        val usesAvailable = itemstack.maxDamage - itemstack.itemDamage
                        val usesToTake = if (usesNeeded > usesAvailable) usesAvailable else usesNeeded

                        stack.itemDamage = stack.itemDamage - usesToTake

                        if (usesNeeded < usesAvailable) {
                            itemstack = itemstack.copy()
                            itemstack!!.itemDamage = itemstack.itemDamage + usesNeeded
                            aitemstack[i] = itemstack
                        }
                    }
                }
            }
        }

        return aitemstack
    }

}
